clients/src/main/java/org/apache/kafka/common/network/NetworkReceive.java

```diff
@@ -107,4 +129,23 @@ public ByteBuffer payload() {
         return this.buffer;
     }
 
+    // We determine if a peer connection uses SSL/TLS by seeing from the first few bytes.
+    private void ensureNotHandshakeFailurePacket(ReadableByteChannel channel, ByteBuffer size) throws IOException {
+        if (ensureThrough)
+            return;
+        ensureThrough = true;
+        String head = Utils.hexToString(size.array());
+        if (SSL_HANDSHAKE_ALERT.matcher(head).find()) {
+            // Actually, SSL record size is 2 bytes, but head byte is already read.
+            ByteBuffer recordSizeBuf = ByteBuffer.allocate(1);
+            channel.read(recordSizeBuf);
```

Comment:
This and the read below. There is no guarantee that either of these non-blocking reads is going to complete here. So the checks would only work depending on which reads complete.